server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java

```diff
@@ -564,234 +423,116 @@ public FlatObjectFieldType fieldType() {
-            subIterators.add(valueAndPathFieldMapper);
-        }
-        if (subIterators.size() == 0) {
-            return super.iterator();
-        }
-        @SuppressWarnings("unchecked")
-        Iterator<Mapper> concat = Iterators.concat(super.iterator(), subIterators.iterator());
-        return concat;
+    protected String contentType() {
+        return CONTENT_TYPE;
     }
 
-    /**
-     * parseValueAddFields method will store data to Lucene.
-     * the JsonToStringXContentParser returns XContentParser with 3 string fields
-     * fieldName, fieldName._value, fieldName._valueAndPath.
-     * parseValueAddFields recognized string by the stringfield name,
-     * fieldName will be store through the parent FlatObjectFieldMapper,which contains all the keys
-     * fieldName._value will be store through the valueFieldMapper, which contains the values of the Json Object
-     * fieldName._valueAndPath will be store through the valueAndPathFieldMapper, which contains the "path=values" format
-     */
-    private void parseValueAddFields(ParseContext context, String value, String fieldName) throws IOException {
+    private void parseObject(XContentParser parser, ParseContext context) throws IOException {
+        assert parser.currentToken() == XContentParser.Token.START_OBJECT;
+        parser.nextToken(); // Skip the outer START_OBJECT. Need to return on END_OBJECT.
 
-        NamedAnalyzer normalizer = fieldType().normalizer();
-        if (normalizer != null) {
-            value = normalizeValue(normalizer, name(), value);
+        LinkedList<String> path = new LinkedList<>(Collections.singleton(fieldType().name()));
+        HashSet<String> leafPaths = new HashSet<>();
+        while (parser.currentToken() != XContentParser.Token.END_OBJECT) {
+            parseToken(parser, context, path, leafPaths);
         }
 
-        String[] valueTypeList = fieldName.split("\\._");
-        String valueType = "._" + valueTypeList[valueTypeList.length - 1];
-
-        if (fieldType.indexOptions() != IndexOptions.NONE || fieldType.stored()) {
-            // convert to utf8 only once before feeding postings/dv/stored fields
-
-            final BytesRef binaryValue = new BytesRef(fieldType().name() + DOT_SYMBOL + value);
-            Field field = new FlatObjectField(fieldType().name(), binaryValue, fieldType);
+        // create FieldNamesField for exist query
+        createFieldNamesField(context, leafPaths);
+    }
 
-            if (fieldType().hasDocValues() == false && fieldType.omitNorms()) {
-                createFieldNamesField(context);
-            }
-            if (fieldName.equals(fieldType().name())) {
-                context.doc().add(field);
-            }
-            if (valueType.equals(VALUE_SUFFIX)) {
-                if (valueFieldMapper != null) {
-                    valueFieldMapper.addField(context, value);
+    private void createFieldNamesField(ParseContext context, HashSet<String> leafPaths) {
```

Comment:
We'll need to be careful around backward compatibility with this.

I think that this is probably the way we should have done the flat object field names in the first place. Unfortunately, we need to make sure that we don't break users who have already created flat object fields.

I think we need to keep the old (not so good) behavior, and guard it behind a check on `context.indexSettings().getIndexVersionCreated()`. For new indices, we can do things the new way, but old indices need to continue to do it the old way.